File: PracticeVectorEmbedding/js/src/MovieService.js
const fs = require("fs");
const path = require("path");
const Movie = require("./model/Movie");
const MovieData = require("./model/MovieData");
const MovieMatch = require("./model/MovieMatch");

class MovieService {
    constructor(embeddingModel, jsonMapper) {
        this.embeddingModel = embeddingModel;
        this.jsonMapper = jsonMapper;
        this.moviesEmbedding = [];
    }

    initializeMovies() {
        const resource = path.join(
            __dirname,
            "../../springDemo/src/main/resources/movies.json"
        );

        const inputStream = fs.readFileSync(resource, "utf8");

        const movieDataList = this.jsonMapper.parse(inputStream).map(
            movieData => new MovieData(
                movieData.title,
                movieData.description
            )
        );

        for (const movieData of movieDataList) {
            const embedding = this.embeddingModel.embed(
                movieData.getDescription()
            );

            const movie = new Movie(
                movieData.getTitle(),
                movieData.getDescription(),
                embedding
            );

            this.moviesEmbedding.push(movie);
        }

        console.log(
            this.moviesEmbedding.length + " movies loaded with embeddings."
        );

        for (const movie of this.moviesEmbedding) {
            console.log(movie.getEmbedding());
        }
    }

    search(query) {
        const userQueryEmbedding = this.embeddingModel.embed(query);

        const matches = [];

        for (const movie of this.moviesEmbedding) {
            const similarity = this.cosineSimilarity(
                userQueryEmbedding,
                movie.getEmbedding()
            );

            const match = new MovieMatch(
                movie.getTitle(),
                movie.getDescription(),
                similarity
            );

            matches.push(match);
        }

        this.sortBySimilarity(matches);

        return this.topKMatches(matches, 3);
    }

    similarMovies(title) {
        const selectedMovie = this.findMovie(title);
        const matches = [];

        for (const movie of this.moviesEmbedding) {
            if (movie.getTitle().toLowerCase() === title.toLowerCase()) {
                continue;
            }

            const similarity = this.cosineSimilarity(
                selectedMovie.getEmbedding(),
                movie.getEmbedding()
            );

            const match = new MovieMatch(
                movie.getTitle(),
                movie.getDescription(),
                similarity
            );

            matches.push(match);
        }

        this.sortBySimilarity(matches);
        return this.topKMatches(matches, 3);
    }

    findMovie(title) {
        for (const movie of this.moviesEmbedding) {
            if (movie.getTitle().toLowerCase() === title.toLowerCase()) {
                return movie;
            }
        }

        throw new Error("Movie not found: " + title);
    }

    cosineSimilarity(a, b) {
        let dotProduct = 0.0;
        let normA = 0.0;
        let normB = 0.0;

        for (let i = 0; i < a.length; i++) {
            dotProduct += a[i] * b[i];
            normA += a[i] * a[i];
            normB += b[i] * b[i];
        }

        if (normA === 0 || normB === 0) {
            return 0.0;
        }

        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }

    sortBySimilarity(matches) {
        matches.sort(
            (first, second) => second.getMatch() - first.getMatch()
        );
    }

    topKMatches(matches, limit) {
        const topMatches = [];

        const numberOfMatches = Math.min(limit, matches.length);

        for (let i = 0; i < numberOfMatches; i++) {
            topMatches.push(matches[i]);
        }

        return topMatches;
    }
}

module.exports = MovieService;

File: PracticeVectorEmbedding/SpringBoot/src/main/java/in/Harsh/demo/MovieService.java
package in.Harsh.demo;

import in.Harsh.demo.model.MovieMatch;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.json.JsonMapper;
import in.Harsh.demo.model.Movie;
import in.Harsh.demo.model.MovieData;
import jakarta.annotation.PostConstruct;
import org.springframework.ai.embedding.EmbeddingModel;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Service;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.json.JsonMapper;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Service
public class MovieService {

    private final EmbeddingModel embeddingModel;
    private final JsonMapper jsonMapper;

    private final List<Movie> moviesEmbedding = new ArrayList<>();

    public MovieService(EmbeddingModel embeddingModel, JsonMapper jsonMapper) {
        this.embeddingModel = embeddingModel;
        this.jsonMapper = jsonMapper;
    }

    @PostConstruct
    public void initializeMovies() throws IOException {

        ClassPathResource resource =
                new ClassPathResource("movies.json");

        InputStream inputStream =
                resource.getInputStream();

        List<MovieData> movieDataList = jsonMapper.readValue(
                inputStream, new TypeReference<>() {});

        for (MovieData movieData : movieDataList) {
            float[] embedding = embeddingModel.embed(
                    movieData.getDescription());

            Movie movie = new Movie(
                    movieData.getTitle(),
                    movieData.getDescription(),
                    embedding
            );

            moviesEmbedding.add(movie);
        }
        inputStream.close();

        System.out.println(
                moviesEmbedding.size() + " movies loaded with embeddings."
        );

        for(Movie movie : moviesEmbedding) {
            System.out.println(Arrays.toString(movie.getEmbedding()));
        }

    }

    public List<MovieMatch> search(String query) {
        float[] userQueryEmbedding = embeddingModel.embed(query);

        List<MovieMatch> matches = new ArrayList<>();

        for(Movie movie : moviesEmbedding) {
            double similarity = cosineSimilarity(userQueryEmbedding, movie.getEmbedding());

            MovieMatch match =
                    new MovieMatch(movie.getTitle(), movie.getDescription(), similarity);

            matches.add(match);
        }

        sortBySimilarity(matches);

        return topKMatches(matches, 3);
    }

    public List<MovieMatch> similarMovies(String title) {
        Movie selectedMovie = findMovie(title);
        List<MovieMatch> matches = new ArrayList<>();

        for (Movie movie : moviesEmbedding) {
            if (movie.getTitle().equalsIgnoreCase(title)) {
                continue;
            }

            double similarity = cosineSimilarity(
                    selectedMovie.getEmbedding(),
                    movie.getEmbedding());

            MovieMatch match = new MovieMatch(
                    movie.getTitle(),
                    movie.getDescription(),
                    similarity);

            matches.add(match);
        }
        sortBySimilarity(matches);
        return topKMatches(matches, 3);
    }

    private Movie findMovie(String title) {
        for (Movie movie : moviesEmbedding) {
            if (movie.getTitle().equalsIgnoreCase(title)) {
                return movie;
            }
        }

        throw new IllegalArgumentException(
                "Movie not found: " + title
        );
    }

    private double cosineSimilarity(float[] a, float[] b) {
        double dotProduct = 0.0;
        double normA = 0.0;
        double normB = 0.0;

        for (int i = 0; i < a.length; i++) {
            dotProduct += (a[i] * b[i]);
            normA += (a[i] * a[i]);
            normB += (b[i] * b[i]);
        }

        if (normA == 0 || normB == 0) {
            return 0.0;
        }

        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }

    private void sortBySimilarity(List<MovieMatch> matches) {
        Collections.sort(matches,
                (first, second) -> Double.compare(
                        second.getMatch(),
                        first.getMatch())
        );
    }

    private List<MovieMatch> topKMatches(
            List<MovieMatch> matches, int limit) {

        List<MovieMatch> topMatches = new ArrayList<>();

        int numberOfMatches = Math.min(limit, matches.size());

        for (int i = 0; i < numberOfMatches; i++) {
            topMatches.add(matches.get(i));
        }

        return topMatches;
    }
}
